src/openrct2/scripting/ScriptEngine.cpp

```diff
@@ -1127,6 +1134,44 @@ void ScriptEngine::SaveSharedStorage()
     }
 }
 
+#    ifndef DISABLE_NETWORK
+void ScriptEngine::AddSocket(const std::shared_ptr<ScSocketBase>& socket)
+{
+    _sockets.push_back(socket);
+}
+#    endif
+
+void ScriptEngine::UpdateSockets()
+{
+#    ifndef DISABLE_NETWORK
+    // Use simple for i loop as Update calls can modify the list
+    auto it = _sockets.begin();
+    while (it != _sockets.end())
+    {
+        auto& socket = *it;
+        socket->Update();
+        if (socket->IsDisposed())
+        {
+            it = _sockets.erase(it);
+        }
+        else
+        {
+            it++;
+        }
+    }
+#    endif
+}
+
+void ScriptEngine::RemoveSockets(const std::shared_ptr<Plugin>& plugin)
+{
+#    ifndef DISABLE_NETWORK
+    _sockets.erase(
+        std::remove_if(
+            _sockets.begin(), _sockets.end(), [&plugin](const auto& socket) { return socket->GetPlugin() == plugin; }),
```

Comment:
It doesn't look like this condition will match, so sockets are never closed.

Reply:
I tested this method and it works.

Reply:
When I leave a game and return to the main menu listening sockets aren't closed.

Reply:
For some reason something, probably dukglue / duktape is holding on to the ScSocket reference. I have therefore changed back to explicitly closing the socket when the plugin is stopped.

Reply:
I'll checkout this commit and test it out. Thanks for the quick response.